fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java

```diff
@@ -138,6 +141,74 @@ public LuceneSearchProvider(String luceneIndexPath, int maxResults) {
         this(luceneIndexPath, maxResults, LUCENE_SEARCH_TIMEOUT_MS);
     }
 
+    /**
+     * Attempts to acquire the read lock. If waiting for the read lock times out or waiting for the lock
+     * an exception is thrown and the query is failed. Timeout is equivalent to the timeout on a search. If a
+     * read lock waits that long then the query is likely going to timeout anyways.
+     */
+    protected void readLock() {
+        try {
+            if (!lock.readLock().tryLock(searchTimeout, TimeUnit.MILLISECONDS)) {
+                String msg = String.format(
+                        ErrorMessageFormat.LUCENE_LOCK_TIMEOUT.getMessageFormat(),
+                        getDimension().getApiName()
+                );
+
+                LOG.error(msg);
+                throw new IllegalStateException(msg);
+            }
+        } catch (InterruptedException e) {
+            String msg = String.format(
+                    ErrorMessageFormat.LUCENE_LOCK_INTERRUPTED.getMessageFormat(),
+                    getDimension()
+            );
+            LOG.error(msg);
+            throw new IllegalStateException(msg, e);
+        }
+    }
+
+    /**
+     * Unlocks the held read lock for this thread.
+     */
+    protected void readUnlock() {
+        lock.readLock().unlock();
+    }
+
+    /**
+     * Attempts to acquire the write lock. If waiting for the read lock times out or waiting for the lock
+     * an exception is thrown and the query is failed. Timeout is equivalent to the timeout on a search. If a
+     * write lock waits that long then the query is likely going to timeout anyways.
+     */
+    protected void writeLock() {
+        try {
+            // write lock can wait for 1.2 query timeout length before replace occurs, to let current queries drain.
+            if (!lock.writeLock().tryLock((int) (searchTimeout * 1.2), TimeUnit.MILLISECONDS)) {
```

Comment:
We should probably make the multiplier a config variable. This seems like something that could differ based on a customer's setup, and we don't really want to force them to extend LuceneSearchProvider and duplicate this function just to change one number.